link@rel=stylesheet: Don’t load if HTTP status isn’t an OK status

sideshowbarker opened the pull request:
#### d84fa2dfd249fe54b862bbf673c466263c0c5fc8
<pre>
link@rel=stylesheet: Don’t load if HTTP status isn’t an OK status
<a href="https://bugs.webkit.org/show_bug.cgi?id=262030">https://bugs.webkit.org/show_bug.cgi?id=262030</a>

Reviewed by Chris Dumez.

When the HTTP response status code for a &lt;link rel=stylesheet&gt; resource
is a non-OK status (not in the range 200–299), this change ensures
WebKit fully conforms to the requirement in the HTML standard at
<a href="https://html.spec.whatwg.org/#fetching-and-processing-a-resource-from-a-link-element">https://html.spec.whatwg.org/#fetching-and-processing-a-resource-from-a-link-element</a>:processresponseconsumebody
that no “load” event is fired at the element.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/document-metadata/the-link-element/resources/302-no-Location-header-text-css.asis: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/document-metadata/the-link-element/stylesheet-non-OK-status-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/document-metadata/the-link-element/stylesheet-non-OK-status.html: Added.
* Source/WebCore/css/StyleSheetContents.cpp:
(WebCore::StyleSheetContents::parseAuthorStyleSheet):
* Source/WebCore/loader/cache/CachedCSSStyleSheet.cpp:
(WebCore::CachedCSSStyleSheet::sheetText const):
(WebCore::CachedCSSStyleSheet::canUseSheet const):
* Source/WebCore/loader/cache/CachedCSSStyleSheet.h:

Canonical link: <a href="https://commits.webkit.org/268779@main">https://commits.webkit.org/268779@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0fc5ed2363721af1e5e7744f455608fa64af95b2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/20622 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/21040 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/21694 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/22514 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/19249 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/20859 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/24294 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/21220 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/20593 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/20844 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/20676 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/17927 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/23371 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/17842 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/18734 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/25003 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/18921 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/18911 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/22946 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/19499 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/16566 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/18717 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/18570 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4957 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/23054 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/19316 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/loader/cache/CachedCSSStyleSheet.cpp:
```diff
@@ -139,11 +139,18 @@ bool CachedCSSStyleSheet::mimeTypeAllowedByNosniff() const
         return false;
 
+    // https://html.spec.whatwg.org/#fetching-and-processing-a-resource-from-a-link-element:processresponseconsumebody
+    if (response().isNull() && !(response().httpStatusCode() >= 200 && response().httpStatusCode() <= 299)) {
```
You can call `!response.isSuccessful()` to achieve the same thing.

sideshowbarker replied:
Thanks — made it so

cdumez commented on Source/WebCore/loader/cache/CachedCSSStyleSheet.cpp:
```diff
@@ -139,11 +139,18 @@ bool CachedCSSStyleSheet::mimeTypeAllowedByNosniff() const
         return false;
 
+    // https://html.spec.whatwg.org/#fetching-and-processing-a-resource-from-a-link-element:processresponseconsumebody
+    if (response().httpStatusCode() && !response().isSuccessful()) {
```
I find the `response().httpStatusCode()` check here confusing given that we're not using the status code explicitly in the second condition. Would a `!response().isNull()` work? If so, I suggest we switch to that.

sideshowbarker replied:
Thanks — it does seems like `! response.inNull()` would work, but I'll test it locally to make sure, and then push an update

sideshowbarker replied:
> I find the `response().httpStatusCode()` check here confusing given that we're not using the status code explicitly in the second condition. Would a `!response().isNull()` work? If so, I suggest we switch to that.

OK, it seems `!response().isNull()` won’t work, because even for `file:` URLs, `response().isNull()` is not true — so `!response().isNull() && !response().isSuccessful()` becomes effectively the same thing as just checking `!response().isSuccessful()`, which for `file:` URLs will always be true.

So I’ve switched the code back to using `response().httpStatusCode() && !response().isSuccessful()`, but also added a comment to clarify what it’s doing and why (that `response().httpStatusCode()` is essentially all about dealing with the `file:` URL case).

I tried to make it even more clear with an explicit `response().httpStatusCode() != 0` — but the style linter doesn’t like that (message: _“Tests for true/false, null/non-null, and zero/non-zero should all be done without equality comparisons”_), so I reverted it to just `response().httpStatusCode() && !response().isSuccessful()`.

cdumez replied:
Could we use `response().url().protocolIsInHTTPFamily() && !response().isSuccessful()` then?

sideshowbarker replied:
> Could we use `response().url().protocolIsInHTTPFamily() && !response().isSuccessful()` then?

Aha, didn’t know about that one — will try it right now

cdumez reviewed (approved):
LGTM with comment.

cdumez reviewed (approved):
LGTM

sideshowbarker committed:
link@rel=stylesheet: Don’t load if HTTP status isn’t an OK status
https://bugs.webkit.org/show_bug.cgi?id=262030

Reviewed by Chris Dumez.

When the HTTP response status code for a <link rel=stylesheet> resource
is a non-OK status (not in the range 200–299), this change ensures
WebKit fully conforms to the requirement in the HTML standard at
https://html.spec.whatwg.org/#fetching-and-processing-a-resource-from-a-link-element:processresponseconsumebody
that no “load” event is fired at the element.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/document-metadata/the-link-element/resources/302-no-Location-header-text-css.asis: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/document-metadata/the-link-element/stylesheet-non-OK-status-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/document-metadata/the-link-element/stylesheet-non-OK-status.html: Added.
* Source/WebCore/css/StyleSheetContents.cpp:
(WebCore::StyleSheetContents::parseAuthorStyleSheet):
* Source/WebCore/loader/cache/CachedCSSStyleSheet.cpp:
(WebCore::CachedCSSStyleSheet::sheetText const):
(WebCore::CachedCSSStyleSheet::canUseSheet const):
* Source/WebCore/loader/cache/CachedCSSStyleSheet.h:

Canonical link: https://commits.webkit.org/268779@main
